presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -189,6 +210,49 @@ private void enforceTimeLimits()
         }
     }
 
+    /**
+     *  When cluster reaches max tasks limit and also a single query
+     *  excceds a threshold,  kill this query
+     */
+    private void enforceTaskLimits()
+    {
+        int totalRunningTaskCount = 0;
+        int highestRunningTaskCount = 0;
+        Optional<T> highestRunningTaskQuery = Optional.empty();
+        for (T query : queries.values()) {
+            if (query.isDone()) {
+                continue;
+            }
+            int runningTaskCount = query.getRunningTask();
+            totalRunningTaskCount += runningTaskCount;
+            if (runningTaskCount > highestRunningTaskCount) {
+                highestRunningTaskCount = runningTaskCount;
+                highestRunningTaskQuery = Optional.of(query);
+            }
+        }
+
+        runningTaskCount.set(totalRunningTaskCount);
+
+        if (totalRunningTaskCount > maxTotalRunningTaskCount &&
+                highestRunningTaskCount > maxQueryRunningTaskCount &&
+                highestRunningTaskQuery.isPresent()) {
+            highestRunningTaskQuery.get().fail(new PrestoException(QUERY_HAS_TOO_MANY_STAGES, format(
+                    "Cluster is overloaded with too many tasks (%s), query with highest running tasks is killed (%s). " + TOO_MANY_STAGES_MESSAGE,
+                    totalRunningTaskCount, highestRunningTaskCount)));
+            queriesKilledDueToTooManyTask.incrementAndGet();
+        }
+    }
+
+    public long getRunningTaskCount()
```

Comment:
Public and private methods are interleaved. Is that something we do in Presto? Do we have a specific ordering of public/private/package private and static things as well as nested classes that we follow?